### DISCUSSION TITLE
Assigning Environments to Custom Domains

### DISCUSSION THREAD

**Discussion by AndrewBarba** [vercel/vercel#4062] [General] [answered]
The new [Environments](https://zeit.co/blog/environment-variables-ui) feature is a great step towards managing production/staging environments under a single project, but I think it can be taken one step further.

When using the git integration, it's implied that the production environment is only used for the "default" branch. Problem is, our default branch is actually the staging branch, and we deploy to prod when merging to master.

We have the ability to assign domains to specific git branches, but we have no way of controlling which environment is used on that branch.

What I'd love to be able to accomplish is the following:

Domain | Branch | Env | Notes
------- | ------- | ----- | ----
www.example.com | master | Production | The primary domain
beta.example.com | staging | Production | Used internally by non-devs to test new features with prod environment
dev.example.com | staging | Preview | Used only by developers

I could see how assigning multiple domains to a single branch, with different environments, could pose issues for Zeit's deployment model where a domain is simply assigned to a deployment. This would effectively require two builds to be made for the same commit.

One alternative that would allow us to accomplish the same thing: allow connecting two projects to the same git project. We would still need to be able to assign environments to domains, but would not need ability to assign multiple domains to same branch. Then we could create a web-prod and web-staging project and connect both of them to git. Assuming deploy hooks fire to both projects, we would accomplish running two deploys for the same commit.

**Comment by mcsdevv** [comment, marked as answer]
Hey @AndrewBarba, thanks for providing such detailed feedback, much appreciated! We've been gathering feedback on use cases and the 'staging' environment is something that we are looking at currently. As the multiple environments support progresses further we'll be sure to take this into account and find ways to support more use cases by providing greater flexibility.

**Reply by morgoe** [reply]
@msweeneydev does this mean that, at the moment, I can't assign all branches except `master` on Github to push a branch using the `preview` variables? They have to use the `production` variables, unless I deploy from the command line?

**Reply by mcsdevv** [reply]
@morgoe, the default (commonly "master") branch, will receive Production environment variables. All other branches will receive Preview environment variables.

**Reply by mjurincic** [reply]
I think a perfect place for this would be project level Settings -> Domains -> Edit Modal, just add key - value pair input boxes, underneath Git branch selection, that way you get maximum flexibility to override variables based on custom Domain/Git-Branch mapping. If you need more environments create new custom domain/subdomain and make specific overrides.

**Reply by AndrewBarba** [reply]
The correct place for this is definitely Settings -> Domains but I feel strongly you should not allow custom key/value pairs for each domain. Reason being, you will need to create a deployment for every unique combination of key value pairs. Domains are simply aliases for a deployment so can't just alias a domain with different key/value pairs to a deployment, they need to be baked on build. To limit this surface area, I think simply selecting an environment for each domain is good enough. This breaks that barrier of only the "default" branch getting production environment and still allows for a lot of flexibility. In the future I could see Zeit allowing us to create new Environments and then this would work right along side that.
